velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -434,6 +436,136 @@ TpchPlan TpchQueryBuilder::getQ6Plan() const {
+              "",
+              {"supp_nation", "c_custkey"})
+          .hashJoin(
+              {"c_custkey"},
+              {"o_custkey"},
+              orders,
+              "",
+              {"supp_nation", "o_orderkey"})
+          .planNode();
+
+  auto nation_2 =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kNation_2, nationSelectedRowType_2, nationFileColumns_2, {}, nationFilter)
+          .capturePlanNodeId(nationScanNodeId_2)
+          .project({"n_name as cust_nation", "n_nationkey as n2_nationkey"})
+          .planNode();
+
+  auto lineitem =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kLineitem, lineitemSelectedRowType, lineitemFileColumns, {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeId)
+          .project({"l_extendedprice * (1.0 - l_discount) as part_revenue", "l_orderkey", "l_suppkey", "l_shipdate"})
+          .planNode();
+
+  auto supplier = PlanBuilder(planNodeIdGenerator)
+                  .tableScan(kSupplier, supplierSelectedRowType, supplierFileColumns)
+                  .capturePlanNodeId(supplierScanNodeId)
+                  .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kLineitem, lineitemSelectedRowType, lineitemFileColumns, {shipDateFilter})
+          .capturePlanNodeId(lineitemScanNodeId)
+          .project({"l_extendedprice * (1.0 - l_discount) as part_revenue", "l_orderkey", "l_suppkey", "year(l_shipdate) as l_year"})
+          .hashJoin(
+              {"l_suppkey"},
+              {"s_suppkey"},
+              supplier,
+              "",
+              {"s_nationkey", "l_orderkey", "part_revenue", "l_year"})
+          .hashJoin(
+              {"s_nationkey"},
+              {"n2_nationkey"},
+              nation_2,
+              "",
+              {"cust_nation", "part_revenue", "l_orderkey", "l_year"})
+          .hashJoin(
+              {"l_orderkey"},
+              {"o_orderkey"},
+              plan_1,
+              "(((cust_nation = 'FRANCE') AND (supp_nation = 'GERMANY')) OR "
+              "((cust_nation = 'GERMANY') AND (supp_nation = 'FRANCE')))",
+              {"cust_nation", "supp_nation", "part_revenue", "l_year"})
+          .partialAggregation({"supp_nation", "cust_nation", "l_year"}, {"sum(part_revenue) as revenue"})
+          .localPartition({})
+          .finalAggregation()
+          .orderBy({"supp_nation", "cust_nation", "l_year"}, false)
+          .project({"supp_nation", "cust_nation", "l_year", "revenue"})
+          .planNode();
+
+  TpchPlan context;
+  context.plan = std::move(plan);
+  context.dataFiles[customerScanNodeId] = getTableFilePaths(kCustomer);
+  context.dataFiles[ordersScanNodeId] = getTableFilePaths(kOrders);
+  context.dataFiles[lineitemScanNodeId] = getTableFilePaths(kLineitem);
+  context.dataFiles[supplierScanNodeId] = getTableFilePaths(kSupplier);
+  context.dataFiles[nationScanNodeId_1] = getTableFilePaths(kNation);
```

Comment:
For the tables that are used more than once in the same query, you have to bind the scanNode each time in the dataFiles here. e.g. You have to add here 

context.dataFiles[nationScanNodeId_2] = getTableFilePaths(kNation);

The previous query was hanging because the table in the second usage didn't get any splits.

Please can you add so and then try the plans.

Reply:
Thanks for sharing, I will modify accordingly and update the query plan.